python/cudf_polars/cudf_polars/dsl/to_ast.py

```diff
@@ -126,13 +132,95 @@ class ExprTransformerState(TypedDict):
     table_ref: plc.expressions.TableReference
 
 
+ValidateTransformer: TypeAlias = GenericTransformer[expr.Expr, None, ValidateState]
 Transformer: TypeAlias = GenericTransformer[expr.Expr, plc_expr.Expression, ASTState]
 ExprTransformer: TypeAlias = GenericTransformer[
     expr.Expr, expr.Expr, ExprTransformerState
 ]
 """Protocol for transformation of Expr nodes."""
 
 
+@singledispatch
+def _validate_to_ast(node: expr.Expr, self: ValidateTransformer) -> None:
+    print(f"Unhandled expression type {type(node)}")
+    raise NotImplementedError(f"Unhandled expression type {type(node)}")
+
+
+@_validate_to_ast.register
+def _(node: expr.Literal, self: ValidateTransformer) -> None:
+    return None
+
+
+@_validate_to_ast.register
+def _(node: expr.Col, self: ValidateTransformer) -> None:
+    if self.state["for_parquet"]:
+        return None
+    raise TypeError("Should always be wrapped in a ColRef node before translation")
+
+
+@_validate_to_ast.register
+def _(node: expr.ColRef, self: ValidateTransformer) -> None:
+    if self.state["for_parquet"]:
+        raise TypeError("Not expecting ColRef node in parquet filter")
+
+
+@_validate_to_ast.register
+def _(node: expr.BinOp, self: ValidateTransformer) -> None:
+    if node.op == plc.binaryop.BinaryOperator.NULL_NOT_EQUALS:
+        return None
+    if self.state["for_parquet"]:
+        op1_col, op2_col = (isinstance(op, expr.Col) for op in node.children)
+        if op1_col ^ op2_col:
+            op: plc.binaryop.BinaryOperator = node.op
+            if op not in SUPPORTED_STATISTICS_BINOPS:
+                raise NotImplementedError(
+                    f"Parquet filter binop with column doesn't support {node.op!r}"
+                )
+            op1, op2 = node.children
+            if op2_col:
+                (op1, op2) = (op2, op1)
+                op = REVERSED_COMPARISON[op]
+            if not isinstance(op2, expr.Literal):
+                raise NotImplementedError(
+                    "Parquet filter binops must have form 'col binop literal'"
+                )
+            return None
+        elif op1_col and op2_col:
+            raise NotImplementedError(
+                "Parquet filter binops must have one column reference not two"
+            )
+    for child in node.children:
+        self(child)
+
+
+@_validate_to_ast.register
+def _(node: expr.BooleanFunction, self: ValidateTransformer) -> None:
+    if node.name is expr.BooleanFunction.Name.IsIn:
+        needles, haystack = node.children
+        if isinstance(haystack, expr.LiteralColumn) and len(haystack.value) < 16:
+            return None
```

Comment:
I've dropped a call to `self(needles)` here, which should probably be kept (in case it raises).